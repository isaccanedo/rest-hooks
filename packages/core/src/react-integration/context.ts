import { createContext } from 'react';
import type { DenormalizeCache } from '@rest-hooks/normalizr';

import type { ActionTypes, State } from '../types.js';
import { initialState } from '../state/createReducer.js';
import Controller from '../controller/Controller.js';

export const StateContext = createContext(initialState);

const dispatch = (value: ActionTypes) => {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    console.error(
      'It appears you are trying to use Rest Hooks without a provider.\nFollow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component',
    );
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'test') {
      console.error(
        'If you are trying to test: https://resthooks.io/docs/guides/unit-testing-hooks',
      );
    }
  }
  return Promise.resolve();
};
export const DispatchContext = createContext(dispatch);

// this is not needed anymore, but keeping around for backcompatibility
export const DenormalizeCacheContext = createContext<DenormalizeCache>({
  entities: {},
  results: {},
});

export const ControllerContext = createContext<Controller>(
  new Controller({
    dispatch,
    globalCache: {
      entities: {},
      results: {},
    },
  }),
);

export interface Store<S> {
  subscribe(listener: () => void): () => void;
  dispatch: React.Dispatch<ActionTypes>;
  getState(): S;
  uninitialized?: boolean;
}
/* istanbul ignore next */
export const StoreContext = createContext<Store<State<unknown>>>({
  subscribe: listener => () => {},
  dispatch: () => {},
  getState: () => initialState,
  uninitialized: true,
});
