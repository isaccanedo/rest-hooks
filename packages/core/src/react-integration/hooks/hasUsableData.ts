import { FetchShape } from '../../endpoint/index.js';

/** @deprecated use https://resthooks.io/docs/api/Controller#getResponse */
export default function hasUsableData(
  fetchShape: Pick<FetchShape<any>, 'options'>,
  cacheReady: boolean,
  deleted: boolean,
  invalidated?: boolean,
) {
  return (
    !deleted &&
    !fetchShape.options?.invalidIfStale &&
    cacheReady &&
    !invalidated
  );
}
