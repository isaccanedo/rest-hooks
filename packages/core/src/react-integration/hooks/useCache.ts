/* eslint-disable @typescript-eslint/ban-types */
import { useMemo } from 'react';
import {
  EndpointInterface,
  DenormalizeNullable,
  Schema,
  FetchFunction,
  ResolveType,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useCacheNew from '../newhooks/useCache.js';

/**
 * Access a response if it is available.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://resthooks.io/docs/api/useCache
 */
export default function useCache<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(
  endpoint: E,
  ...args: Args
): E['schema'] extends {}
  ? DenormalizeNullable<E['schema']>
  : E extends (...args: any) => any
  ? ResolveType<E> | undefined
  : any {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useCacheNew(adaptedEndpoint, ...args);
}
