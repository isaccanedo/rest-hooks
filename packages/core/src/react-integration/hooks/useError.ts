import type { NetworkError, UnknownError } from '@rest-hooks/normalizr';
import { useMemo } from 'react';
import {
  EndpointInterface,
  Schema,
  FetchFunction,
} from '@rest-hooks/normalizr';

import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import shapeToEndpoint from '../../endpoint/adapter.js';
import useErrorNew from '../newhooks/useError.js';

export type ErrorTypes = NetworkError | UnknownError;

type UseErrorReturn<P> = P extends [null] ? undefined : ErrorTypes | undefined;

/**
 * Get any errors for a given request
 * @see https://resthooks.io/docs/api/useError
 */
export default function useError<
  E extends
    | Pick<
        EndpointInterface<FetchFunction, Schema | undefined, undefined>,
        'key' | 'schema' | 'invalidIfStale'
      >
    | Pick<ReadShape<any, any>, 'getFetchKey' | 'schema' | 'options'>,
  Args extends
    | (E extends { key: any }
        ? readonly [...Parameters<E['key']>]
        : readonly [ParamsFromShape<E>])
    | readonly [null],
>(endpoint: E, ...args: Args): UseErrorReturn<typeof args> {
  const adaptedEndpoint: any = useMemo(() => {
    return shapeToEndpoint(endpoint);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return useErrorNew(adaptedEndpoint, ...args);
}
