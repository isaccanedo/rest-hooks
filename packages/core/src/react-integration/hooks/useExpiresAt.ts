import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import useMeta from './useMeta.js';

/** Returns whether the data at this url is fresh or stale
 * @deprecated use https://resthooks.io/docs/api/Controller#getResponse
 */
export default function useExpiresAt<
  Shape extends Pick<ReadShape<any, any>, 'getFetchKey' | 'options'>,
>(
  fetchShape: Shape,
  params: ParamsFromShape<Shape> | null,
  entitiesExpireAt = 0,
): number {
  const meta = useMeta(fetchShape, params);

  if (!meta) {
    return entitiesExpireAt;
  }

  return meta.expiresAt;
}
