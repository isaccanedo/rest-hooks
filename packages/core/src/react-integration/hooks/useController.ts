import { useContext } from 'react';

import Controller from '../../controller/Controller.js';
import { ControllerContext } from '../context.js';

/**
 * Imperative control of Rest Hooks store
 * @see https://resthooks.io/docs/api/useController
 */
export default function useController(): Controller {
  return useContext(ControllerContext);
}
