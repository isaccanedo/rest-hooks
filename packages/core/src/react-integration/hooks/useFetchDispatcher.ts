import { Schema } from '@rest-hooks/normalizr';
import { useContext, useCallback } from 'react';

import {
  FetchShape,
  SchemaFromShape,
  ParamsFromShape,
  BodyFromShape,
  OptimisticUpdateParams,
  ReturnFromShape,
} from '../../endpoint/index.js';
import { DispatchContext } from '../context.js';
import createFetch from '../../state/actions/createFetch.js';

/** Build an imperative dispatcher to issue network requests.
 * @deprecated use https://resthooks.io/docs/api/Controller#fetch
 */
export default function useFetchDispatcher(throttle = false): <
  Shape extends FetchShape<Schema, Readonly<object>, any>,
  UpdateParams extends OptimisticUpdateParams<
    SchemaFromShape<Shape>,
    FetchShape<any, any, any>
  >[],
>(
  fetchShape: Shape & {
    update?: (...args: any) => Record<string, (...args: any) => any>;
  },
  params: ParamsFromShape<Shape>,
  body: BodyFromShape<Shape>,
  updateParams?: UpdateParams | undefined,
) => ReturnFromShape<typeof fetchShape> {
  const dispatch = useContext(DispatchContext);

  const fetchDispatcher = useCallback(
    <Shape extends FetchShape<Schema, Readonly<object>, any>>(
      fetchShape: Shape,
      params: ParamsFromShape<Shape>,
      body: BodyFromShape<Shape>,
      updateParams?:
        | OptimisticUpdateParams<
            SchemaFromShape<Shape>,
            FetchShape<any, any, any>
          >[]
        | undefined,
    ) => {
      const action = createFetch(fetchShape, {
        params,
        body,
        throttle,
        updateParams,
      });
      dispatch(action);
      return action.meta.promise;
    },
    [dispatch, throttle],
  );
  // any is due to the ternary that we don't want to deal with in our implementation
  return fetchDispatcher as any;
}
