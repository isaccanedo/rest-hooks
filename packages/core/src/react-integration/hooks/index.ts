import useFetcher from './useFetcher.js';
import useController from './useController.js';
import useCache from './useCache.js';
import useRetrieve from './useRetrieve.js';
import useResource from './useResource.js';
import useSubscription from './useSubscription.js';
import useMeta from './useMeta.js';
import useError from './useError.js';
import useExpiresAt from './useExpiresAt.js';
import useInvalidator from './useInvalidator.js';
import useResetter from './useResetter.js';
import useFetchDispatcher from './useFetchDispatcher.js';
import useInvalidateDispatcher from './useInvalidateDispatcher.js';
import useFetch from '../newhooks/useFetch.js';
import useSuspense from '../newhooks/useSuspense.js';
export { default as hasUsableData } from './hasUsableData.js';
export type { ErrorTypes } from './useError.js';

export {
  useFetcher,
  useController,
  useFetchDispatcher,
  useCache,
  useError,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useExpiresAt,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  useFetch,
  useSuspense,
};
