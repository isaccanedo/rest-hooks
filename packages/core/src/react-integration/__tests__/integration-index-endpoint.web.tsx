import {
  CoolerArticleResource,
  ArticleResource,
  PaginatedArticleResource,
  UserResource,
  ArticleResourceWithOtherListUrl,
  ListPaginatedArticle,
  CoolerArticleDetail,
  TypedArticleResource,
  IndexedUserResource,
  UnionResource,
  FutureArticleResource,
  User,
} from '__tests__/new';
import nock from 'nock';
import { act } from '@testing-library/react-hooks';
import { useContext } from 'react';

// relative imports to avoid circular dependency in tsconfig references
import {
  makeRenderRestHook,
  makeCacheProvider,
  makeExternalCacheProvider,
} from '../../../../test';
import { useCache, useSuspense, useController } from '../hooks';
import {
  payload,
  createPayload,
  users,
  nested,
  paginatedFirstPage,
  paginatedSecondPage,
  valuesFixture,
} from '../test-fixtures';
import { StateContext } from '../context';

function onError(e: any) {
  e.preventDefault();
}
beforeEach(() => {
  if (typeof addEventListener === 'function')
    addEventListener('error', onError);
});
afterEach(() => {
  if (typeof removeEventListener === 'function')
    removeEventListener('error', onError);
});

describe('indexes', () => {
  let renderRestHook: ReturnType<typeof makeRenderRestHook>;
  let mynock: nock.Scope;

  beforeEach(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload)
      .delete(`/article-cooler/${payload.id}`)
      .reply(204, '')
      .delete(`/article/${payload.id}`)
      .reply(200, {})
      .delete(`/user/23`)
      .reply(204, '')
      .get(`/article-cooler/0`)
      .reply(403, {})
      .get(`/article-cooler/666`)
      .reply(200, '')
      .get(`/article-cooler/`)
      .reply(200, nested)
      .get(`/article-cooler/values`)
      .reply(200, valuesFixture)
      .post(`/article-cooler/`)
      .reply(200, createPayload)
      .get(`/user`)
      .reply(200, users);

    mynock = nock(/.*/).defaultReplyHeaders({
      'Access-Control-Allow-Origin': '*',
      'Content-Type': 'application/json',
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  beforeEach(() => {
    renderRestHook = makeRenderRestHook(makeCacheProvider);
  });

  it('should resolve parallel useSuspense() request', async () => {
    const { result, waitForNextUpdate } = renderRestHook(() => {
      const { fetch } = useController();
      useSuspense(IndexedUserResource.getList);
      return {
        bob: useCache(IndexedUserResource.getIndex, {
          username: 'bob',
        }),
        charlie: useCache(IndexedUserResource.getIndex, {
          username: 'charlie',
        }),
        fetch,
        state: useContext(StateContext),
      };
    });
    expect(result.current).toBeUndefined();
    await waitForNextUpdate();
    const bob = result.current.bob;
    expect(bob).toBeDefined();
    expect(bob instanceof User).toBe(true);
    expect(bob?.username).toBe('bob');
    expect(bob).toMatchSnapshot();
    expect(result.current.charlie).toBeUndefined();

    const renamed = { ...users[0] };
    renamed.username = 'charlie';
    mynock.get(`/user/23`).reply(200, renamed);
    result.current.fetch(IndexedUserResource.get, { id: '23' });
    await waitForNextUpdate();

    const charlie = result.current.charlie;
    expect(charlie).toBeDefined();
    expect(charlie instanceof User).toBe(true);
    expect(charlie?.username).toBe('charlie');
    expect(charlie).toMatchSnapshot();
    expect(result.current.bob).toBeUndefined();

    // deletes should roll through to indexes
    await act(async () => {
      await result.current.fetch(IndexedUserResource.delete, { id: '23' });
    });
    expect(result.current.charlie).toBeUndefined();
  });
});
