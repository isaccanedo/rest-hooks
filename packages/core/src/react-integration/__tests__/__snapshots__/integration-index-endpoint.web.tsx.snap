// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexes should resolve parallel useSuspense() request 1`] = `
IndexedUser {
  "email": "bob@bob.com",
  "id": 23,
  "isAdmin": false,
  "username": "bob",
}
`;

exports[`indexes should resolve parallel useSuspense() request 2`] = `
IndexedUser {
  "email": "bob@bob.com",
  "id": 23,
  "isAdmin": false,
  "username": "charlie",
}
`;
